# CSE 151A Group Project

In [81]:
import pandas as pd
import datetime as dt

## Data Exploration

Start by loading in the dataset. Since the reviews csv doesn't provide anything aside from the fact that a certain location id got reviewed by a certain user id, we opted to not use it. 

In [82]:
listings_df = pd.read_csv('Listings.csv', encoding='ISO-8859-1')
listings_df.head()

c:\Users\Home\anaconda3\envs\dsc80\lib\site-packages\IPython\core\interactiveshell.py:3508: DtypeWarning: Columns (5,13) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


listing_id                                              name   host_id  \
0      281420    Beautiful Flat in le Village Montmartre, Paris   1466919   
1     3705183                   39 m² Paris (Sacre CÃ
“ur)  10328771   
2     4082273               Lovely apartment with Terrace, 60m2  19252768   
3     4797344               Cosy studio (close to Eiffel tower)  10668311   
4     4823489  Close to Eiffel Tower - Beautiful flat : 2 rooms  24837558   

   host_since                 host_location host_response_time  \
0  2011-12-03  Paris, Ile-de-France, France                NaN   
1  2013-11-29  Paris, Ile-de-France, France                NaN   
2  2014-07-31  Paris, Ile-de-France, France                NaN   
3  2013-12-17  Paris, Ile-de-France, France                NaN   
4  2014-12-14  Paris, Ile-de-France, France                NaN   

   host_response_rate  host_acceptance_rate host_is_superhost  \
0                 NaN                   NaN                 f   
1                 NaN                   NaN                 f   
2                 NaN                   NaN                 f   
3                 NaN                   NaN                 f   
4                 NaN                   NaN                 f   

   host_total_listings_count  ... minimum_nights maximum_nights  \
0                        1.0  ...              2           1125   
1                        1.0  ...              2           1125   
2                        1.0  ...              2           1125   
3                        1.0  ...              2           1125   
4                        1.0  ...              2           1125   

  review_scores_rating review_scores_accuracy review_scores_cleanliness  \
0                100.0                   10.0                      10.0   
1                100.0                   10.0                      10.0   
2                100.0                   10.0                      10.0   
3                100.0                   10.0                      10.0   
4                100.0                   10.0                      10.0   

   review_scores_checkin  review_scores_communication review_scores_location  \
0                   10.0                         10.0                   10.0   
1                   10.0                         10.0                   10.0   
2                   10.0                         10.0                   10.0   
3                   10.0                         10.0                   10.0   
4                   10.0                         10.0                   10.0   

  review_scores_value  instant_bookable  
0                10.0                 f  
1                10.0                 f  
2                10.0                 f  
3                10.0                 f  
4                10.0                 f  

[5 rows x 33 columns]

Let's start by looking at the shape of the dataset.

In [83]:
listings_df.shape

(279712, 33)

We have 279712 locations, each with 33 variables associated with them. 

Let's start by taking a look at what each column means.

In [84]:
listings_info = pd.read_csv('Listings_data_dictionary.csv', encoding='ISO-8859-1')
listings_info

Field  \
0                    listing_id   
1                          name   
2                       host_id   
3                    host_since   
4                 host_location   
5            host_response_time   
6            host_response_rate   
7          host_acceptance_rate   
8             host_is_superhost   
9     host_total_listings_count   
10         host_has_profile_pic   
11       host_identity_verified   
12                neighbourhood   
13                     district   
14                         city   
15                     latitude   
16                    longitude   
17                property_type   
18                    room_type   
19                 accommodates   
20                     bedrooms   
21                    amenities   
22                        price   
23               minimum_nights   
24               maximum_nights   
25         review_scores_rating   
26       review_scores_accuracy   
27    review_scores_cleanliness   
28        review_scores_checkin   
29  review_scores_communication   
30       review_scores_location   
31          review_scores_value   
32             instant_bookable   

                                          Description  
0                                          Listing ID  
1                                        Listing Name  
2                                             Host ID  
3                         Date the Host joined Airbnb  
4                    Location where the Host is based  
5      Estimate of how long the Host takes to respond  
6               Percentage of times the Host responds  
7   Percentage of times the Host accepts a booking...  
8   Binary field to determine if the Host is a Sup...  
9               Total listings the Host has in Airbnb  
10  Binary field to determine if the Host has a pr...  
11  Binary field to determine if the Host has a ve...  
12                     Neighborhood the Listing is in  
13                         District the Listing is in  
14                             City the Listing is in  
15                                 Listing's latitude  
16                                Listing's longitude  
17                   Type of property for the Listing  
18        Type of room type in Airbnb for the Listing  
19                     Guests the Listing accomodates  
20                            Bedrooms in the Listing  
21                     Amenities the Listing includes  
22         Listing price (in each country's currency)  
23                         Minimum nights per booking  
24                         Maximum nights per booking  
25              Listing's overall rating (out of 100)  
26  Listing's accuracy score based on what's promo...  
27            Listing's cleanliness score (out of 10)  
28    Listing's check-in experience score (out of 10)  
29  Listing's communication with the Host score (o...  
30  Listing's location score within the city (out ...  
31  Listing's value score relative to its price (o...  
32  Binary field to determine if the Listing can b...

Now that we have a look at what each column means, let's start by dropping every column that couldn't possibly be used to predict the rating of the listing.

We'll also drop any columns that are associated with the rating, because that information is not available at the time of booking.

In [85]:
listings_df.columns

Index(['listing_id', 'name', 'host_id', 'host_since', 'host_location',
       'host_response_time', 'host_response_rate', 'host_acceptance_rate',
       'host_is_superhost', 'host_total_listings_count',
       'host_has_profile_pic', 'host_identity_verified', 'neighbourhood',
       'district', 'city', 'latitude', 'longitude', 'property_type',
       'room_type', 'accommodates', 'bedrooms', 'amenities', 'price',
       'minimum_nights', 'maximum_nights', 'review_scores_rating',
       'review_scores_accuracy', 'review_scores_cleanliness',
       'review_scores_checkin', 'review_scores_communication',
       'review_scores_location', 'review_scores_value', 'instant_bookable'],
      dtype='object')

In [86]:
unusable_cols = ["listing_id","name","host_id"]
unavailable_cols = ['review_scores_accuracy', 'review_scores_cleanliness',
                    'review_scores_checkin', 'review_scores_communication',
                    'review_scores_location', 'review_scores_value']

listings_df = listings_df.drop(unusable_cols,axis=1)
listings_df = listings_df.drop(unavailable_cols,axis=1)
listings_df.columns

Index(['host_since', 'host_location', 'host_response_time',
       'host_response_rate', 'host_acceptance_rate', 'host_is_superhost',
       'host_total_listings_count', 'host_has_profile_pic',
       'host_identity_verified', 'neighbourhood', 'district', 'city',
       'latitude', 'longitude', 'property_type', 'room_type', 'accommodates',
       'bedrooms', 'amenities', 'price', 'minimum_nights', 'maximum_nights',
       'review_scores_rating', 'instant_bookable'],
      dtype='object')

Next, we're going to take a look at the missing values in the dataset.

In [87]:
for column in listings_df.columns:
    print(column + ": " + str(listings_df[column].isna().sum()) + " missing values")

host_since: 165 missing values
host_location: 840 missing values
host_response_time: 128782 missing values
host_response_rate: 128782 missing values
host_acceptance_rate: 113087 missing values
host_is_superhost: 165 missing values
host_total_listings_count: 165 missing values
host_has_profile_pic: 165 missing values
host_identity_verified: 165 missing values
neighbourhood: 0 missing values
district: 242700 missing values
city: 0 missing values
latitude: 0 missing values
longitude: 0 missing values
property_type: 0 missing values
room_type: 0 missing values
accommodates: 0 missing values
bedrooms: 29435 missing values
amenities: 0 missing values
price: 0 missing values
minimum_nights: 0 missing values
maximum_nights: 0 missing values
review_scores_rating: 91405 missing values
instant_bookable: 0 missing values


The first thing we noticed was that many of the host columns had the same number of missing values, which we guessed meant that AirBNB simply didn't have information on those hosts. Let's verify this.

In [88]:
missing_165 = ['host_since', 'host_is_superhost', 'host_total_listings_count', "host_has_profile_pic", "host_identity_verified"]
missing_count = listings_df[listings_df[missing_165].isna().all(axis=1)].shape[0]

print(missing_count == 165)

True


Great. We're going to delete all rows that have missing data in all of these columns, as we think that sacrificing such a small fraction of the dataset for additional variables is worthwhile.

In [89]:
listings_df = listings_df[listings_df[missing_165].isna().all(axis=1) == False]
listings_df.shape

(279547, 24)

We also decided that due to the amount of missing values in host response times, host response rates, district, and host acceptance rates, we should omit those variables.

In [90]:
listings_df = listings_df.drop(columns=['host_response_time','host_response_rate', 'host_acceptance_rate', "district"])

Bedrooms was a tough decision given the amount of missing data, but we opted to drop all rows where the number of bedrooms was missing, considering that it'd be tough to make a decision on booking an AirBNB without its number of bedrooms. More on this later.

In [91]:
listings_df = listings_df[listings_df["bedrooms"].isna() == False]

We decided latitude and longitude would be redundant when city is already provided, so we dropped that variable. We'll also drop neighborhood, as we feel it's unlikely to have enough data points for each neighborhood to be of use.

In [92]:
listings_df = listings_df.drop(columns=["latitude","longitude", "neighbourhood"])

We'll also drop host_location, as we don't see it being effective in predicting rating.

In [93]:
listings_df = listings_df.drop(columns=["host_location"])

Finally, we're also going to have to drop all rows where review_scores_rating is null, as this is the variable we need to predict.

In [94]:
listings_df = listings_df[listings_df["review_scores_rating"].isna() == False]

Great, now we have eliminated all of our missing values.

In [95]:
for column in listings_df.columns:
    print(column + ": " + str(listings_df[column].isna().sum()) + " missing values")

host_since: 0 missing values
host_is_superhost: 0 missing values
host_total_listings_count: 0 missing values
host_has_profile_pic: 0 missing values
host_identity_verified: 0 missing values
city: 0 missing values
property_type: 0 missing values
room_type: 0 missing values
accommodates: 0 missing values
bedrooms: 0 missing values
amenities: 0 missing values
price: 0 missing values
minimum_nights: 0 missing values
maximum_nights: 0 missing values
review_scores_rating: 0 missing values
instant_bookable: 0 missing values


In [96]:
listings_df.shape

(168414, 16)

We're now down to 168414 entries with 17 variables. Let's walk through the design rationale for each column:

- host_since -- we want the amount of time a host has been active, as we think this would've given them time to adjust their properties to be of higher quality.
- All other host variables -- gives us a decent idea of the quality of the host.
- city -- gives us a good idea of their rough location.
- property_type and room_type -- the type of housing may affect how well they're generally rated
- accomodates / bedrooms -- the size of the AirBNB might follow general trends in rating
- amenities -- gives us the added benefits, which may add a convenience that'll boost rating.
- price -- rating may be correlated with pricing, one way or another
- minimum_nights and maximum_nights -- we think that this'll add to or subtract from the convenience for the user, leading to correlation with rating
- instant_bookable -- same as above.

Now, let's begin by evaluating for preprocessing.

In [97]:
listings_df.head()

host_since host_is_superhost  host_total_listings_count  \
0  2011-12-03                 f                        1.0   
1  2013-11-29                 f                        1.0   
2  2014-07-31                 f                        1.0   
3  2013-12-17                 f                        1.0   
4  2014-12-14                 f                        1.0   

  host_has_profile_pic host_identity_verified   city     property_type  \
0                    t                      f  Paris  Entire apartment   
1                    t                      t  Paris  Entire apartment   
2                    t                      f  Paris  Entire apartment   
3                    t                      t  Paris  Entire apartment   
4                    t                      f  Paris  Entire apartment   

      room_type  accommodates  bedrooms  \
0  Entire place             2       1.0   
1  Entire place             2       1.0   
2  Entire place             2       1.0   
3  Entire place             2       1.0   
4  Entire place             2       1.0   

                                           amenities  price  minimum_nights  \
0  ["Heating", "Kitchen", "Washer", "Wifi", "Long...     53               2   
1  ["Shampoo", "Heating", "Kitchen", "Essentials"...    120               2   
2  ["Heating", "TV", "Kitchen", "Washer", "Wifi",...     89               2   
3  ["Heating", "TV", "Kitchen", "Wifi", "Long ter...     58               2   
4  ["Heating", "TV", "Kitchen", "Essentials", "Ha...     60               2   

   maximum_nights  review_scores_rating instant_bookable  
0            1125                 100.0                f  
1            1125                 100.0                f  
2            1125                 100.0                f  
3            1125                 100.0                f  
4            1125                 100.0                f

First, since we want host_since to represent the amount of time the host has been active, we'll convert from datetime to a straight integer by subtracting it from the current date.

In [98]:
listings_df["host_since"] = pd.to_datetime(listings_df["host_since"])
listings_df["host_since"] = listings_df["host_since"].apply(lambda x: dt.datetime.now() - x)
listings_df["host_since"] = listings_df["host_since"].apply(lambda x: x.days)

Next, we'll convert all columns consisting of t's and f's to 1's and 0's such that it can be more easily used by ML models.

In [99]:
tf_cols = ["host_is_superhost", "host_has_profile_pic", "host_identity_verified", "instant_bookable"]
listings_df[tf_cols] = listings_df[tf_cols].replace({'t': 1, 'f': 0})

Next, we'll one-hot encode all of the categorical variables with the exception of amenities.

In [100]:
cat_cols = ["property_type", "room_type", "city"]
listings_df = pd.get_dummies(listings_df, columns=cat_cols)